WTF::ThreadAssertion is not useful as most of the code is run with a WorkQueue

kkinnunen-apple opened the pull request:
#### 8ed46d3dd638db662ea8093cec940b1b00992ecf
<pre>
WTF::ThreadAssertion is not useful as most of the code is run with a WorkQueue
<a href="https://bugs.webkit.org/show_bug.cgi?id=237026">https://bugs.webkit.org/show_bug.cgi?id=237026</a>
rdar://problem/89608226

Reviewed by Chris Dumez.

Rename ThreadAssertion to ThreadlikeAssertion, and assert that if the assertion is created in a WorkQueue,
the assertion checks for the work queue. If it is created on a Thread, it checks for that.

Implement for non-Cocoa by just using thread id for the sequence id. All WorkQueues have unique threads
and all Threads have unique ids. This means WorkQueues can be distinquished from Threads without WorkQueues.

Implement for Cocoa by using the same uid id namespace for WorkQueues as Threads have. Store the uid
in a Dispatch dispatch queue specific field. If the code is not running in Dispatch, return the current
thread uid. Since the uid namespace is the same, WorkQueues can be distinuished from Threads without WorkQueues.

In both cases, non-WebKit platform threads get also a Thread, so they&apos;re distinquished from WorkQueues.

Currently it is common in WebKit code to use a pattern of:
ASSERT(RunLoop::isMain()) -- This is not run in a work queue or a specific thread
ASSERT(!RunLoop::isMain()) -- This is run in a specific thread, but because we use WorkQueues,
we cannot actually assert the specific thread.

This pattern could be for example used to assert that object is created and destroyed in
correct thread.

After this commit, that can be expressed more exactly as:
 NO_UNIQUE_ADDRESS ThreadlikeAssertion m_owner;

* Source/WTF/wtf/MainThread.h:
(WTF::WTF_ASSERTS_ACQUIRED_CAPABILITY):
* Source/WTF/wtf/ThreadAssertions.h:
(WTF::WTF_ASSERTS_ACQUIRED_CAPABILITY):
(WTF::Threadlike::createThreadlikeAssertion):
* Source/WTF/wtf/Threading.cpp:
(WTF::Threadlike::currentSequence):
* Source/WTF/wtf/Threading.h:
(WTF::WTF_ASSERTS_ACQUIRED_CAPABILITY):
(WTF::Thread::Thread): Deleted.
* Source/WTF/wtf/WorkQueue.h:
(WTF::WTF_ASSERTS_ACQUIRED_CAPABILITY):
* Source/WTF/wtf/cocoa/WorkQueueCocoa.cpp:
(WTF::WorkQueueBase::platformInitialize):
(WTF::WorkQueue::WorkQueue):
(WTF::WorkQueue::threadlikeAssertion const):
(WTF::WorkQueue::assertIsCurrent const): Deleted.
* Source/WTF/wtf/generic/WorkQueueGeneric.cpp:
(WTF::WorkQueueBase::threadlikeAssertion const):
(WTF::WorkQueue::assertIsCurrent const): Deleted.
* Source/WebCore/platform/graphics/mac/DisplayConfigurationMonitor.cpp:
* Source/WebKit/GPUProcess/media/RemoteVideoFrameObjectHeap.cpp:
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayer.h:
* Tools/TestWebKitAPI/Tests/WTF/ThreadAssertionsTest.cpp:
* Tools/TestWebKitAPI/Tests/WTF/WorkQueue.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/255274@main">https://commits.webkit.org/255274@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7d40a78d5a2e85ac07c316fa338becec18f76d3a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/91951 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1195 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22486 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/101774 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/161833 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1191 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/29638 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/84425 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/97980 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97609 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/729 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/78502 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27690 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82651 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/82253 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/70726 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/83312 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36043 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/16271 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/78368 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33780 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17372 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/27047 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3653 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37657 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/40086 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/80990 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39541 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36494 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17781 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/ThreadAssertions.h:
```diff
@@ -25,13 +25,22 @@
 
+class ThreadlikeAssertion;
+
+class Threadlike {
```
I think "thread-like" is two words so I would have chosen "ThreadLike" instead of "Threadlike".

kkinnunen-apple replied:
done

kkinnunen-apple committed:
WTF::ThreadAssertion is not useful as most of the code is run with a WorkQueue
https://bugs.webkit.org/show_bug.cgi?id=237026
rdar://problem/89608226

Reviewed by Chris Dumez.

Rename ThreadAssertion to ThreadlikeAssertion, and assert that if the assertion is created in a WorkQueue,
the assertion checks for the work queue. If it is created on a Thread, it checks for that.

Implement for non-Cocoa by just using thread id for the sequence id. All WorkQueues have unique threads
and all Threads have unique ids. This means WorkQueues can be distinquished from Threads without WorkQueues.

Implement for Cocoa by using the same uid id namespace for WorkQueues as Threads have. Store the uid
in a Dispatch dispatch queue specific field. If the code is not running in Dispatch, return the current
thread uid. Since the uid namespace is the same, WorkQueues can be distinuished from Threads without WorkQueues.

In both cases, non-WebKit platform threads get also a Thread, so they're distinquished from WorkQueues.

Currently it is common in WebKit code to use a pattern of:
ASSERT(RunLoop::isMain()) -- This is not run in a work queue or a specific thread
ASSERT(!RunLoop::isMain()) -- This is run in a specific thread, but because we use WorkQueues,
we cannot actually assert the specific thread.

This pattern could be for example used to assert that object is created and destroyed in
correct thread.

After this commit, that can be expressed more exactly as:
 NO_UNIQUE_ADDRESS ThreadlikeAssertion m_owner;

* Source/WTF/wtf/MainThread.h:
(WTF::WTF_ASSERTS_ACQUIRED_CAPABILITY):
* Source/WTF/wtf/ThreadAssertions.h:
(WTF::WTF_ASSERTS_ACQUIRED_CAPABILITY):
(WTF::Threadlike::createThreadlikeAssertion):
* Source/WTF/wtf/Threading.cpp:
(WTF::Threadlike::currentSequence):
* Source/WTF/wtf/Threading.h:
(WTF::WTF_ASSERTS_ACQUIRED_CAPABILITY):
(WTF::Thread::Thread): Deleted.
* Source/WTF/wtf/WorkQueue.h:
(WTF::WTF_ASSERTS_ACQUIRED_CAPABILITY):
* Source/WTF/wtf/cocoa/WorkQueueCocoa.cpp:
(WTF::WorkQueueBase::platformInitialize):
(WTF::WorkQueue::WorkQueue):
(WTF::WorkQueue::threadlikeAssertion const):
(WTF::WorkQueue::assertIsCurrent const): Deleted.
* Source/WTF/wtf/generic/WorkQueueGeneric.cpp:
(WTF::WorkQueueBase::threadlikeAssertion const):
(WTF::WorkQueue::assertIsCurrent const): Deleted.
* Source/WebCore/platform/graphics/mac/DisplayConfigurationMonitor.cpp:
* Source/WebKit/GPUProcess/media/RemoteVideoFrameObjectHeap.cpp:
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayer.h:
* Tools/TestWebKitAPI/Tests/WTF/ThreadAssertionsTest.cpp:
* Tools/TestWebKitAPI/Tests/WTF/WorkQueue.cpp:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/255274@main
